metricflow/dataflow/builder/dataflow_plan_builder.py

```diff
@@ -225,23 +228,57 @@ def _build_derived_metric_output_node(
             f"For {metric.type} metric: {metric_spec}, needed metrics are:\n"
             f"{pformat_big_objects(metric_input_specs=metric_input_specs)}"
         )
-        return ComputeMetricsNode(
-            parent_node=self._build_metrics_output_node(
-                metric_specs=metric_input_specs,
-                queried_linkable_specs=queried_linkable_specs,
-                where_constraint=where_constraint,
+
+        parent_nodes: List[BaseOutput] = []
+        metric_has_time_offset = bool(metric_spec.offset_window or metric_spec.offset_to_grain)
+        for metric_input_spec in metric_input_specs:
+            parent_nodes.append(
+                self._build_any_metric_output_node(
+                    metric_spec=MetricSpec(
+                        element_name=metric_input_spec.element_name,
+                        constraint=metric_input_spec.constraint,
+                        alias=metric_input_spec.alias,
+                        offset_window=metric_input_spec.offset_window,
+                        offset_to_grain=metric_input_spec.offset_to_grain,
+                    ),
+                    queried_linkable_specs=queried_linkable_specs,
+                    # If metric is offset, we'll apply constraint after offset to avoid removing values unexpectedly.
+                    where_constraint=where_constraint if not metric_has_time_offset else None,
+                    time_range_constraint=time_range_constraint if not metric_has_time_offset else None,
+                )
+            )
+
+        parent_node = parent_nodes[0] if len(parent_nodes) == 1 else CombineMetricsNode(parent_nodes=parent_nodes)
+        output_node: BaseOutput = ComputeMetricsNode(parent_node=parent_node, metric_specs=[metric_spec])
+
+        # For nested ratio / derived metrics with time offset, apply offset & where constraint after metric computation.
```

Comment:
This method got bigger in a hurry but I don't have any better suggestions since we wire so many parameters around.